Description: Add basic build support for loongarch64.
Author: Dandan Zhang <zhangdandan@loongson.cn>
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1057643
Reviewed-by: Shengqi Chen <harry@debian.org>
Last-Update: 2025-04-07
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
Note: support for loong64 is experimental and could be removed without
prior notification.
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -70,6 +70,8 @@
         set(DNNL_TARGET_ARCH "S390X")
     elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(rv.*|RV.*|riscv.*|RISCV.*)")
         set(DNNL_TARGET_ARCH "RV64")
+    elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(loongarch64.*|LOONGARCH64.*|loong64.*|LOONG64.*)")
+        set(DNNL_TARGET_ARCH "LOONGARCH64")
     else()
         set(DNNL_TARGET_ARCH "X64")
     endif()
--- a/cmake/platform.cmake
+++ b/cmake/platform.cmake
@@ -329,6 +329,8 @@
         elseif(DNNL_TARGET_ARCH STREQUAL "RV64")
              # G = General-purpose extensions, C = Compression extension (very common).
              append(DEF_ARCH_OPT_FLAGS "${RV64_MARCH_FLAG}")
+        elseif(DNNL_TARGET_ARCH STREQUAL "LOONGARCH64")
+             append(DEF_ARCH_OPT_FLAGS "-march=loongarch64")
         elseif(DNNL_TARGET_ARCH STREQUAL "X64")
              platform_clang_x64_arch_ccxx_flags(DEF_ARCH_OPT_FLAGS)
         endif()
@@ -450,6 +452,8 @@
         elseif(DNNL_TARGET_ARCH STREQUAL "RV64")
             # G = General-purpose extensions, C = Compression extension (very common).
             append(DEF_ARCH_OPT_FLAGS "${RV64_MARCH_FLAG}")
+        elseif(DNNL_TARGET_ARCH STREQUAL "LOONGARCH64")
+             append(DEF_ARCH_OPT_FLAGS "-march=loongarch64")
         elseif(DNNL_TARGET_ARCH STREQUAL "X64")
             platform_gnu_x64_arch_ccxx_flags(DEF_ARCH_OPT_FLAGS)
         endif()
--- a/src/cpu/README.md
+++ b/src/cpu/README.md
@@ -48,6 +48,7 @@
 - `DNNL_PPC64` is 1 on OpenPOWER / IBM Power architecture;
 - `DNNL_S390X` is 1 on IBMz / s390x architecture;
 - `DNNL_RV64` is 1 on RISC-V architecture;
+- `DNNL_LOONGARCH64` is 1 on LoongArch architecture;
 - `DNNL_ARCH_GENERIC` is 1 on other platforms.
 Only one of the macros above is defined to 1. All others are defined to 0.
 
--- a/src/cpu/platform.hpp
+++ b/src/cpu/platform.hpp
@@ -30,11 +30,12 @@
 // - DNNL_PPC64
 // - DNNL_S390X
 // - DNNL_RV64
+// - DNNL_LOONGARCH64
 // - DNNL_ARCH_GENERIC
 // Target architecture macro is set to 1, others to 0. All macros are defined.
 
 #if defined(DNNL_X64) + defined(DNNL_AARCH64) + defined(DNNL_PPC64) \
-                + defined(DNNL_S390X) + defined(DNNL_RV64) \
+                + defined(DNNL_S390X) + defined(DNNL_RV64) + defined(DNNL_LOONGARCH64) \
                 + defined(DNNL_ARCH_GENERIC) \
         == 0
 #if defined(__x86_64__) || defined(_M_X64)
@@ -47,13 +48,15 @@
 #define DNNL_S390X 1
 #elif defined(__riscv)
 #define DNNL_RV64 1
+#elif defined(__loongarch_lp64)
+#define DNNL_LOONGARCH64 1
 #else
 #define DNNL_ARCH_GENERIC 1
 #endif
 #endif // defined(DNNL_X64) + ... == 0
 
 #if defined(DNNL_X64) + defined(DNNL_AARCH64) + defined(DNNL_PPC64) \
-                + defined(DNNL_S390X) + defined(DNNL_RV64) \
+                + defined(DNNL_S390X) + defined(DNNL_RV64) + defined(DNNL_LOONGARCH64) \
                 + defined(DNNL_ARCH_GENERIC) \
         != 1
 #error One and only one architecture should be defined at a time
@@ -74,6 +77,9 @@
 #if !defined(DNNL_RV64)
 #define DNNL_RV64 0
 #endif
+#if !defined(DNNL_LOONGARCH64)
+#define DNNL_LOONGARCH64 0
+#endif
 #if !defined(DNNL_ARCH_GENERIC)
 #define DNNL_ARCH_GENERIC 0
 #endif
